### PR TITLE
fix: handle 127 error code for podman compatibility

### PR DESCRIPTION
**Pull request by vchandela**
## What does this PR do?

- Handle 127 error code for podman compatibility.
- After merging this, we still need to disable ryuk for tests to work with Podman. Raised an issue for this: https://github.com/testcontainers/testcontainers-go/issues/2781

## Why is it important?

- Docker has a bug where it overrides exit code = 127 with exit code = 126. However, Podman handles these error codes separately.
- While testing etcd module, my container was returning exit code = 127 on Podman. test-containers.go didn't know what to do with 127 error code as it follows the moby project so, it was stuck in an infinite loop.
- Although, the container was up but the code was panicking.

## Related issues

- Some relevant PRs in other projects:
  - https://github.com/containers/podman/issues/367
  - https://github.com/moby/moby/issues/45795

## How to test this PR

- Output for `podman exec -ti <containerid> /bin/sh -c "echo test"; echo $?`

```
podman exec -ti 3604e691a9efbaa916bdeaa365f7481a04aca79722c430d3cc285b671215cfec /bin/sh -c "echo test"; echo $?
Error: crun: executable file `/bin/sh` not found in $PATH: No such file or directory: OCI runtime attempted to invoke a command that was not found
127
```

### CODE REVIEW

File: wait/host_port.go [hunk 1]
```diff
@@ -209,7 +209,7 @@ func internalCheck(ctx context.Context, internalPort nat.Port, target StrategyTa
 
 		if exitCode == 0 {
 			break
-		} else if exitCode == 126 {
+		} else if exitCode == 126 || exitCode == 127 {
```
Comment (stevenh) [inline wait/host_port.go:212]: suggestion: Could we add a comment about why we have multiple values and extract them to constants so they are not magic numbers any more.
126 is also used in `host_port_test.go` too.

You can find the details of what they [represent here](https://github.com/moby/moby/blob/b3c775059314039807862defe9f531c2bca049db/daemon/errors.go#L122).

**Review by stevenh — CHANGES_REQUESTED** [review]

File: wait/host_port.go (hunk 1 above)
Comment (vchandela) [reply wait/host_port.go]: Done.

File: wait/host_port.go [hunk 2]
```diff
@@ -207,10 +215,16 @@ func internalCheck(ctx context.Context, internalPort nat.Port, target StrategyTa
...
+			Docker maps exit code 127 to 126, while Podman treats them separately.
+			Handle both to ensure compatibility with Docker and Podman.
+		*/
 		if exitCode == 0 {
 			break
-		} else if exitCode == 126 {
+		} else if exitCode == exitEaccess {
 			return errShellNotExecutable
+		} else if exitCode == exitCmdNotFound {
+			return errShellNotFound
```
Comment (stevenh) [inline wait/host_port.go:227]: suggestion: For multiple cases like this a select is easer to follow. The return nil at the end would need to be removed too.

This won't apply cleanly because github doesn't allow suggestions to unedited code:
```suggestion
		select exitCode {
		case 0:
			return nil
		case exitEaccess:
			return errShellNotExecutable
		case exitCmdNotFound:
			return errShellNotFound
		}
```

Comment (stevenh) [inline wait/host_port.go:227]: This would also need to be handled by callers, so we'll need additional updates.

File: wait/host_port.go [hunk 3]
```diff
@@ -13,13 +13,21 @@ import (
 	"github.com/docker/go-connections/nat"
 )
 
+type exitStatus = int
+
+const (
+	exitEaccess     exitStatus = 126 // container cmd can't be invoked (permission denied)
+	exitCmdNotFound exitStatus = 127 // container cmd not found/does not exist or invalid bind-mount
+)
```
Comment (stevenh) [inline wait/host_port.go:21]: suggestion: As we're not using the type here because its not our data, this can be removed.
```suggestion
const (
	exitEaccess             = 126 // container cmd can't be invoked (permission denied)
	exitCmdNotFound = 127 // container cmd not found/does not exist or invalid bind-mount
)
```

**Review by stevenh — CHANGES_REQUESTED** [review]
Thanks for these improvements, some more suggestions

File: wait/host_port.go (hunk 2 above)
Comment (vchandela) [reply wait/host_port.go]: Thanks for pointing out.

File: wait/host_port.go (hunk 3 above)
Comment (vchandela) [reply wait/host_port.go:21]: Done.

**Review by stevenh — CHANGES_REQUESTED** [review]
Looking good, we just need to handle the new error on return from the internalCheck function and we should be good

**Comment by vchandela** [comment]
> Looking good, we just need to handle the new error on return from the internalCheck function and we should be good

do we also need to write a test in host_port_test.go for exitCode = 127?

**Comment by mdelapenya** [comment]
> do we also need to write a test in host_port_test.go for exitCode = 127?

Yes please, once added, I think this PR is good to merge, thanks!

File: wait/host_port_test.go [hunk 4]
```diff
@@ -508,3 +508,58 @@ func TestHostPortStrategySucceedsGivenShellIsNotInstalled(t *testing.T) {
 	err = wg.WaitUntilReady(context.Background(), target)
 	require.NoError(t, err)
 }
+
+func TestHostPortStrategySucceedsGivenShellIsNotFound(t *testing.T) {
```
Comment (mdelapenya) [inline wait/host_port_test.go:512]: cool, thanks for this! Do you think it's possible to add a test for the internal function so that we verify the new error is raised?

File: wait/host_port_test.go [hunk 5]
```diff
@@ -508,3 +508,58 @@ func TestHostPortStrategySucceedsGivenShellIsNotInstalled(t *testing.T) {
...
+	listener, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (stevenh) [inline wait/host_port_test.go:516]: suggestion: could we use require for all checks please, just makes the tests much cleaner, for example:
```suggestion
	require.NoError(t, err)
```

File: wait/host_port_test.go [hunk 6]
```diff
@@ -508,3 +508,58 @@ func TestHostPortStrategySucceedsGivenShellIsNotInstalled(t *testing.T) {
...
+	defer listener.Close()
+
+	rawPort := listener.Addr().(*net.TCPAddr).Port
+	port, err := nat.NewPort("tcp", strconv.Itoa(rawPort))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	target := &MockStrategyTarget{
```
Comment (stevenh) [inline wait/host_port_test.go:525]: suggestion: Could we use the new generated mock `mockStrategyTarget` for this, as we're in the process of moving across to it. You can see an [example in this PR](https://github.com/testcontainers/testcontainers-go/pull/2773).

File: wait/host_port_test.go [hunk 7]
```diff
@@ -508,3 +508,58 @@ func TestHostPortStrategySucceedsGivenShellIsNotInstalled(t *testing.T) {
...
+		},
+	}
+
+	wg := NewHostPortStrategy("80").
+		WithStartupTimeout(5 * time.Second).
+		WithPollInterval(100 * time.Millisecond)
+
+	err = wg.WaitUntilReady(context.Background(), target)
+	require.NoError(t, err)
+}
```
Comment (stevenh) [inline wait/host_port_test.go:565]: suggestion: missing new line at the end of file.

**Review by stevenh — CHANGES_REQUESTED** [review]

**Comment by kiview** [comment]
Just to double check, if https://github.com/moby/moby/issues/45795 would get fixed in the upstream, this PR would still be valid, correct?

**Comment by stevenh** [comment]
> Just to double check, if [moby/moby#45795](https://github.com/moby/moby/issues/45795) would get fixed in the upstream, this PR would still be valid, correct?

Technically it should only check 127, as 126 is an access issue which should not happen for `/bin/sh` so it could hide an issue if for some reason `/bin` has the wrong permissions preventing the internal check from being successful.

File: wait/host_port_test.go (hunk 5 above)
Comment (vchandela) [reply wait/host_port_test.go:516]: done. `require` sounds much better for sure.

File: wait/host_port_test.go (hunk 4 above)
Comment (vchandela) [reply wait/host_port_test.go:486]: done.

File: wait/host_port_test.go (hunk 6 above)
Comment (vchandela) [reply wait/host_port_test.go:495]: @stevenh it would be nice if you could modify the newly added tests in your PR itself as you have context of the changes. This will help in keeping similar changes in the same PR. wdyt?

File: wait/host_port_test.go (hunk 7 above)
Comment (vchandela) [reply wait/host_port_test.go]: I've done it for all the test files.
We can also substitute some conditions with `require.Contains()` but maybe we can pick it up in future PRs. Otherwise, this PR will be too bulky.

File: wait/host_port_test.go (hunk 6 above)
Comment (stevenh) [reply wait/host_port_test.go:495]: I can but I'm trying to avoid a moving target and the functionality already exists, its just not in common use yet.

File: docker_files_test.go [hunk 8]
```diff
@@ -19,14 +19,10 @@ func TestCopyFileToContainer(t *testing.T) {
 
 	// copyFileOnCreate {
 	absPath, err := filepath.Abs(filepath.Join(".", "testdata", "hello.sh"))
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.NoError(t, err)
```
Comment (stevenh) [inline docker_files_test.go:22]: suggestion: sorry the intent was to only use require for the new test, all these changes will cause a bunch of conflicts with other PR's which are inflight, so would we revert all of these and just use it for your new test.

Sorry to be be a pain, but this would cause quit a lot of conflicts.

**Review by stevenh — CHANGES_REQUESTED** [review]
Sorry I think there was a miss understanding about my comment to use require, the intent was only for the new test, not to do it globally, which while desired could cause a lot of conflicts with other PRs that are currently in flight.

File: docker_files_test.go (hunk 8 above)
Comment (vchandela) [reply docker_files_test.go]: Cool. Made the changes.

File: wait/host_port_test.go (hunk 6 above)
Comment (vchandela) [reply wait/host_port_test.go:517]: some `{}_test.go` files in `/wait` have package = `wait_test` while others have only `wait`. What convention do we want to follow?
This causes unnecessary confusion as `MockStrategyTarget` is in package `wait` while `mockStrategyTarget` is in `wait_test`.

Comment (vchandela) [reply wait/host_port_test.go:517]: I don't understand how `MockStrategyTarget` will be migrated to `mockStrategyTarget`. Could you share a sample for it?
MockStrategyTarget was quite easy to understand to be honest. I'm not too sure which example you're talking about in this PR: https://github.com/testcontainers/testcontainers-go/pull/2773

Comment (stevenh) [reply wait/host_port_test.go:517]: You can see an example in the wait file tests here https://github.com/testcontainers/testcontainers-go/blob/main/wait/file_test.go

Comment (vchandela) [reply wait/host_port_test.go]: Done.

File: wait/host_port.go [hunk 9]
```diff
@@ -193,7 +203,7 @@ func externalCheck(ctx context.Context, ipAddress string, port nat.Port, target
 	}
 }
 
-func internalCheck(ctx context.Context, internalPort nat.Port, target StrategyTarget) error {
+func InternalCheck(ctx context.Context, internalPort nat.Port, target StrategyTarget) error {
```
Comment (stevenh) [inline wait/host_port.go:206]: bug: this shouldn’t be exported it’s a private method

File: wait/host_port_test.go [hunk 10]
```diff
@@ -1,18 +1,25 @@
-package wait
+package wait_test
```
Comment (stevenh) [inline wait/host_port_test.go:1]: suggestion: no need to make this change. We will do that separately as it’s just adding unnecessary changes

File: wait/host_port_test.go [hunk 11]
```diff
@@ -456,55 +463,166 @@ func TestHostPortStrategyFailsWhileInternalCheckingDueToUnexpectedContainerStatu
...
-	wg := NewHostPortStrategy("80").
+	wg := wait.NewHostPortStrategy("80").
 		WithStartupTimeout(5 * time.Second).
 		WithPollInterval(100 * time.Millisecond)
 
 	err = wg.WaitUntilReady(context.Background(), target)
 	require.NoError(t, err)
 }
+
+func TestInternalCheckFailsGivenShellIsNotInstalled(t *testing.T) {
```
Comment (stevenh) [inline wait/host_port_test.go:544]: These are the exception to the rule 🙃sorry to mess you about. If you undo the test package change they will all just work without making the internal check public

File: wait/host_port_test.go [hunk 12]
```diff
@@ -1,18 +1,25 @@
...
 	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/errdefs"
 	"github.com/docker/go-connections/nat"
 	"github.com/stretchr/testify/require"
 
 	"github.com/testcontainers/testcontainers-go/exec"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+var (
```
Comment (stevenh) [inline wait/host_port_test.go:21]: This looks unrelated

File: wait/host_port_test.go [hunk 13]
```diff
@@ -456,55 +463,166 @@ func TestHostPortStrategyFailsWhileInternalCheckingDueToUnexpectedContainerStatu
...
-			return &types.ContainerState{
-				Running: true,
-			}, nil
+			},
 		},
-		ExecImpl: func(_ context.Context, _ []string, _ ...exec.ProcessOption) (int, io.Reader, error) {
-			// This is the error that would be returned if the shell is not installed.
-			return 126, nil, nil
+	}, nil)
+	target.EXPECT().MappedPort(anyContext, nat.Port("80")).Return(port, nil)
```
Comment (stevenh) [inline wait/host_port_test.go:489]: suggestion: no need to do all the tests that already in the other PR just the new tests your adding

File: wait/host_port_test.go [hunk 14]
```diff
@@ -456,55 +463,166 @@ func TestHostPortStrategyFailsWhileInternalCheckingDueToUnexpectedContainerStatu
...
+					},
+				},
+			},
+		},
+	}, nil)
+	target.EXPECT().MappedPort(anyContext, nat.Port("80")).Return(port, nil)
+	target.EXPECT().State(anyContext).Return(&types.ContainerState{
+		Running: true,
+	}, nil)
+	target.EXPECT().Exec(anyContext, []string{"/bin/sh", "-c", "true && (\n\t\t\t\t\tcat /proc/net/tcp* | awk '{print $2}' | grep -i :0050 ||\n\t\t\t\t\tnc -vz -w 1 localhost 80 ||\n\t\t\t\t\t/bin/sh -c '</dev/tcp/localhost/80'\n\t\t\t\t)\n\t\t\t\t"}).Return(wait.ExitEaccess, nil, nil)
```
Comment (stevenh) [inline wait/host_port_test.go:573]: suggestion: you don’t need to specify the exact args when they are complex like this just use an any of type

**Review by stevenh — CHANGES_REQUESTED** [review]

File: wait/host_port_test.go (hunk 10 above)
Comment (vchandela) [reply wait/host_port_test.go:1]: If we don't do this, then we'll have to export `mockStrategyTarget` as it is `wait_test` while our tests were originally in `wait`. So, I moved `host_port_test.go` to `wait_test` and exported the necessary variables from `host_port.go`.

File: wait/host_port.go (hunk 9 above)
Comment (vchandela) [reply wait/host_port.go:206]: This is needed to avoid exporting `mockStrategyTarget `. I've explained it better here: https://github.com/testcontainers/testcontainers-go/pull/2778#discussion_r1760575891

File: wait/host_port_test.go (hunk 12 above)
Comment (vchandela) [reply wait/host_port_test.go:21]: After moving `host_port_test.go` to `wait_test`, we need to import `wait` package to get `Internal()`, `ErrPortNotFound`, etc

File: wait/host_port_test.go (hunk 13 above)
Comment (vchandela) [reply wait/host_port_test.go:489]: I'll remove this once other conversations are resolved.

File: wait/host_port_test.go (hunk 14 above)
Comment (vchandela) [reply wait/host_port_test.go:573]: makes sense. I'll incorporate it with the other changes.

File: wait/host_port.go (hunk 9 above)
Comment (stevenh) [reply wait/host_port.go:206]: Sorry we can't export an internal function just be support testing.

File: wait/host_port_test.go (hunk 11 above)
Comment (stevenh) [reply wait/host_port_test.go:544]: This could be testing using the end to end instead of direct call to the method.

File: wait/host_port_test.go (hunk 10 above)
Comment (stevenh) [reply wait/host_port_test.go:1]: You never want to export functions to test like this. There are two options, do a end to end which exercises these cases or a seperate file which which is in the `wait` packages.

Comment (stevenh) [reply wait/host_port_test.go:1]: As it uses the logger you can capture the output in the test and assert the content. e.g.
```go
func MyTest(t *testing.T) {
	oldWriter := log.Writer()
	var buf bytes.Buffer
	log.SetWriter(&buf)
	t.Cleanup(func() {
		log.SetWriter(oldWriter)
	})

	// Run the test
	require.Contains(t, buf.String(), "Shell not found in container")
}
```

File: wait/host_port.go (hunk 9 above)
Comment (vchandela) [reply wait/host_port.go:206]: done.

File: wait/host_port_test.go (hunk 10 above)
Comment (vchandela) [reply wait/host_port_test.go:1]: @stevenh made the changes. PTAL. I've merged internal check tests with the existing tests.
Also, I've reverted back to `MockStrategyTarget` as it will lead to unnecessary code changes for importing certain variables and methods. Please handle `MockStrategyTarget` -> `mockStrategyTarget` for this test in your draft PR.

File: wait/host_port_test.go (hunk 11 above)
Comment (vchandela) [reply wait/host_port_test.go]: I've removed these tests and merged them with the existing ones.

**Comment by mdelapenya** [comment]
@vchandela can you run `make lint` so that all files are gofumpt-ed? 🙏

**Comment by vchandela** [comment]
> @vchandela can you run `make lint` so that all files are gofumpt-ed? 🙏

Done. Thanks for pointing out.

**Review by stevenh — APPROVED** [review]
Looks good to me, thank you for all your effort on the back and forward on this most appreciated.

**Comment by stevenh** [comment]
@vchandela can you check into the CI failures to make sure they aren't fall out from this.

**Comment by vchandela** [comment]
> @vchandela can you check into the CI failures to make sure they aren't fall out from this.

They seem unrelated to this PR. One is for K3s and another is for rabbitmq.

**Review by mdelapenya — APPROVED** [review]
LGTM, thanks!
